### PR TITLE
Add Auth. Issue Custom Page type and handling

### PR DESCRIPTION
**Pull request by kingthorin**
As discussed via email.

Based on previous PRs: #3638, #6194, #6232, and #6245.

Signed-off-by: kingthorin <kingthorin@users.noreply.github.com>

### CODE REVIEW

**Comment by kingthorin** [comment]
I’ll sort out the test failure.

File: zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPage.java [hunk 1]
```diff
@@ -117,7 +117,8 @@ enum Type {
          * OTHER is intended to provide an option for scan rule scripts for usages that may not yet
          * have been planned.
          */
-        OTHER(4, "custompages.type.other");
+        OTHER(4, "custompages.type.other"),
+        AUTH_4XX(5, "custompages.type.auth");
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPage.java:121]: Worth adding JavaDoc for the actual status codes it targets.

File: zap/src/main/java/org/zaproxy/zap/extension/pscan/PassiveScanData.java [hunk 2]
```diff
@@ -247,6 +248,28 @@ public boolean isPageOther(HttpMessage msg) {
...
+    public boolean isPageAuthIssue(HttpMessage msg) {
+        if (isCustomPage(msg, CustomPage.Type.OK_200)) {
+            return false;
+        }
+        if (isCustomPage(msg, CustomPage.Type.AUTH_4XX)) {
+            return true;
+        }
+        int statusCode = msg.getResponseHeader().getStatusCode();
+        return Arrays.asList(HttpStatusCode.UNAUTHORIZED, HttpStatusCode.FORBIDDEN)
+                .contains(statusCode);
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/pscan/PassiveScanData.java:270]: Better with ifs, or extract a constant.

File: zap/src/main/java/org/zaproxy/zap/model/Context.java [hunk 3]
```diff
@@ -864,6 +865,10 @@ private boolean statusCodeFallback(HttpMessage msg, CustomPage.Type cpType) {
                 return msg.getResponseHeader().getStatusCode() == HttpStatusCode.NOT_FOUND;
             case OK_200:
                 return msg.getResponseHeader().getStatusCode() == HttpStatusCode.OK;
+            case AUTH_4XX:
+                int statusCode = msg.getResponseHeader().getStatusCode();
+                return Arrays.asList(HttpStatusCode.UNAUTHORIZED, HttpStatusCode.FORBIDDEN)
+                        .contains(statusCode);
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/model/Context.java:871]: Same.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java [hunk 4]
```diff
@@ -1650,6 +1650,53 @@ void shouldCheckPageOtherWithParent(boolean expectedResult) {
...
+        verify(parent).isCustomPage(message, type);
+        verifyNoInteractions(analyser);
+    }
+
+    @Test
+    void isPageAuthIssueShouldReturnTrueIfCustomPageMatches() {
+        // Given
+        CustomPage.Type type = CustomPage.Type.AUTH_4XX;
+        HttpMessage message = new HttpMessage();
+        //        message.getResponseHeader().setStatusCode(200);
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java:1673]: Remove.

**Comment by thc202** [comment]
It has conflicts.

File: zap/src/main/java/org/zaproxy/zap/extension/pscan/PassiveScanData.java (hunk 2 above)
Comment (kingthorin) [reply zap/src/main/java/org/zaproxy/zap/extension/pscan/PassiveScanData.java:270]: A List constant?

Comment (kingthorin) [reply zap/src/main/java/org/zaproxy/zap/extension/pscan/PassiveScanData.java:270]: Thanks, will do

**Comment by kingthorin** [comment]
Conflict and comments addressed.

**Comment by thc202** [comment]
It has conflicts again, sorry.

**Comment by kingthorin** [comment]
I'll tackle that this evening. If I could get a review anyway that'd really help.

File: zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPage.java [hunk 5]
```diff
@@ -117,7 +117,12 @@ enum Type {
          * OTHER is intended to provide an option for scan rule scripts for usages that may not yet
          * have been planned.
          */
-        OTHER(4, "custompages.type.other");
+        OTHER(4, "custompages.type.other"),
+        /**
+         * AUTH_4XX represents conditions that would normally be covered by HTTP 401 or 403 type
+         * responses.
+         */
+        AUTH_4XX(5, "custompages.type.auth");
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPage.java:125]: Add the since tag.

**Comment by kingthorin** [comment]
Conflict & review addressed.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Comment by kingthorin** [comment]
Thanks, nice to get that finished. Will submit a help PR shortly.

**Pull request merged**